datafusion/optimizer/src/unwrap_cast_in_comparison.rs

```diff
@@ -122,10 +122,6 @@ impl ExprRewriter for UnwrapCastExprRewriter {
                 let right = right.as_ref().clone();
                 let left_type = left.get_type(&self.schema);
                 let right_type = right.get_type(&self.schema);
-                // can't get the data type, just return the expr
-                if left_type.is_err() || right_type.is_err() {
-                    return Ok(expr.clone());
-                }
                 // Because the plan has been done the type coercion, the left and right must be equal
                 let left_type = left_type?;
                 let right_type = right_type?;
```

Comment:
I reviewed this again, and I think this is safe to remove. If we have expressions where we cannot determine the types then it indicates that an earlier rule has introduced an invalid expression.

```suggestion
                let left_type = left.get_type(&self.schema)?;
                let right_type = right.get_type(&self.schema)?;
```

Reply:
If we need to throw the error when can't `get_type` of the expr, I think this fix maybe reasonable.

Reply:
cc @alamb @andygrove

Reply:
>  think the change could be slightly larger, replacing lines 123 through 131 just with:

Done in 139f08d66